File: sdk/go/configstore/defaults_exporter.go
package configstore

import (
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
)

const maxDefaultsExporterErrorBytes = 512

// RunDefaultsExporter runs the small command used by an importing application
// to export source-owned parameter defaults. It returns a process exit code:
// zero on success, two for invalid arguments, and one for all other failures.
//
// A typical main passes os.Args[1:], os.Stdout, os.Stderr, its profile defaults
// provider, and the generated binding's EncodeDefaultsArtifact function. When
// --output is omitted, the artifact is written to stdout.
func RunDefaultsExporter[P ~string, T any](
	args []string,
	stdout io.Writer,
	stderr io.Writer,
	provider func(P) (*T, error),
	encoder func(string, *T) ([]byte, error),
) int {
	if stderr == nil {
		return 1
	}
	flags, err := parseDefaultsExporterFlags(args)
	if err != nil {
		writeDefaultsExporterError(stderr, "arguments", err)
		return 2
	}
	if flags.help {
		if stdout == nil {
			return 1
		}
		if _, err := io.WriteString(stdout, defaultsExporterUsage()); err != nil {
			return 1
		}
		return 0
	}
	if provider == nil {
		writeDefaultsExporterError(stderr, "configuration", errors.New("defaults provider is required"))
		return 1
	}
	if encoder == nil {
		writeDefaultsExporterError(stderr, "configuration", errors.New("artifact encoder is required"))
		return 1
	}

	defaults, err := provider(P(flags.profile))
	if err != nil {
		writeDefaultsExporterError(stderr, "load defaults", errors.New("provider failed"))
		return 1
	}
	if defaults == nil {
		writeDefaultsExporterError(stderr, "load defaults", errors.New("provider returned nil"))
		return 1
	}
	artifactData, err := encoder(flags.profile, defaults)
	if err != nil {
		writeDefaultsExporterError(stderr, "encode artifact", errors.New("encoder failed"))
		return 1
	}
	artifact, err := ParseDefaultsArtifact(artifactData)
	if err != nil {
		writeDefaultsExporterError(stderr, "validate encoded artifact", err)
		return 1
	}
	if artifact.Profile != flags.profile {
		writeDefaultsExporterError(stderr, "validate encoded artifact", errors.New("profile does not match --profile"))
		return 1
	}

	if flags.output == "-" {
		if stdout == nil {
			writeDefaultsExporterError(stderr, "write artifact", errors.New("stdout is required when --output is -"))
			return 1
		}
		if _, err := stdout.Write(artifactData); err != nil {
			writeDefaultsExporterError(stderr, "write artifact", err)
			return 1
		}
		return 0
	}
	if err := writeDefaultsArtifactFile(flags.output, artifactData); err != nil {
		writeDefaultsExporterError(stderr, "write artifact", errors.New("output failed"))
		return 1
	}
	return 0
}

type defaultsExporterFlags struct {
	profile string
	output  string
	help    bool
}

func parseDefaultsExporterFlags(args []string) (defaultsExporterFlags, error) {
	values := make(map[string]string, 2)
	help := false
	for index := 0; index < len(args); index++ {
		argument := args[index]
		if argument == "" {
			continue
		}
		if argument == "--help" || argument == "-h" {
			help = true
			continue
		}
		if !strings.HasPrefix(argument, "--") {
			return defaultsExporterFlags{}, errors.New("positional arguments are not supported")
		}
		name, inline, hasInline := strings.Cut(argument, "=")
		if name != "--profile" && name != "--output" {
			return defaultsExporterFlags{}, fmt.Errorf("unknown option %q", boundedDefaultsOption(name))
		}
		if _, duplicate := values[name]; duplicate {
			return defaultsExporterFlags{}, fmt.Errorf("duplicate option %s", name)
		}
		value := inline
		if !hasInline {
			index++
			if index >= len(args) || strings.HasPrefix(args[index], "--") {
				return defaultsExporterFlags{}, fmt.Errorf("%s requires a value", name)
			}
			value = args[index]
		}
		if value == "" {
			return defaultsExporterFlags{}, fmt.Errorf("%s requires a value", name)
		}
		values[name] = value
	}
	if help {
		return defaultsExporterFlags{help: true}, nil
	}
	profile := values["--profile"]
	if !canonicalDefaultsText(profile, false) {
		return defaultsExporterFlags{}, errors.New("--profile must be nonempty and canonical")
	}
	output := values["--output"]
	if output == "" {
		output = "-"
	}
	return defaultsExporterFlags{profile: profile, output: output}, nil
}

func boundedDefaultsOption(option string) string {
	if len(option) <= 80 {
		return option
	}
	return option[:77] + "..."
}

func defaultsExporterUsage() string {
	return "Usage: defaults-exporter --profile <profile> [--output <file|->]\n\n" +
		"Options:\n" +
		"  --profile <profile>  Application defaults profile\n" +
		"  --output <file|->    Artifact destination (default: - for stdout)\n" +
		"  --help               Show this help\n"
}

func writeDefaultsExporterError(stderr io.Writer, operation string, err error) {
	message := fmt.Sprintf("defaults-exporter: %s: %v", operation, err)
	if len(message) > maxDefaultsExporterErrorBytes {
		message = message[:maxDefaultsExporterErrorBytes-3] + "..."
	}
	_, _ = fmt.Fprintln(stderr, message)
}

func writeDefaultsArtifactFile(path string, data []byte) error {
	directory := filepath.Dir(path)
	if err := os.MkdirAll(directory, 0o755); err != nil {
		return err
	}
	mode := os.FileMode(0o600)
	if info, err := os.Stat(path); err == nil {
		mode = info.Mode().Perm()
	} else if !errors.Is(err, os.ErrNotExist) {
		return err
	}
	temporary, err := os.CreateTemp(directory, ".kms-defaults-*")
	if err != nil {
		return err
	}
	temporaryPath := temporary.Name()
	committed := false
	defer func() {
		if !committed {
			_ = os.Remove(temporaryPath)
		}
	}()
	if err := temporary.Chmod(mode); err != nil {
		_ = temporary.Close()
		return err
	}
	if _, err := temporary.Write(data); err != nil {
		_ = temporary.Close()
		return err
	}
	if err := temporary.Sync(); err != nil {
		_ = temporary.Close()
		return err
	}
	if err := temporary.Close(); err != nil {
		return err
	}
	if err := os.Rename(temporaryPath, path); err != nil {
		return err
	}
	committed = true
	return nil
}

File: sdk/typescript/src/configgen/defaults-exporter.ts
import { randomUUID } from "node:crypto";
import { mkdir, open, rename, stat, unlink } from "node:fs/promises";
import { dirname, resolve } from "node:path";

import {
  parseDefaultsArtifact,
  validateDefaultsProfile,
} from "../configstore/defaults-artifact.js";

const MAX_ERROR_BYTES = 512;

export interface DefaultsExporterIO {
  readonly stdout: (message: string) => void;
  readonly stderr: (message: string) => void;
}

export type DefaultsProvider<T, P extends string = string> = (profile: P) => T | Promise<T>;
export type DefaultsEncoder<T, P extends string = string> = (
  profile: P,
  defaults: T,
) => string | Promise<string>;

/**
 * Run a complete defaults exporter. Applications only supply their existing
 * profile provider and generated encoder.
 */
export async function runDefaultsExporter<P extends string, T>(
  args: readonly string[],
  provider: DefaultsProvider<T, P>,
  encoder: DefaultsEncoder<T, P>,
  io: DefaultsExporterIO = {
    stdout: (message) => process.stdout.write(message),
    stderr: (message) => process.stderr.write(message),
  },
): Promise<number> {
  let flags: ExporterFlags;
  try {
    flags = parseExporterFlags(args);
  } catch (error) {
    io.stderr(`${boundedMessage(error, "defaults exporter: invalid arguments")}\n`);
    io.stderr(exporterUsage());
    return 2;
  }
  if (flags.help) {
    io.stdout(exporterUsage());
    return 0;
  }
  if (typeof provider !== "function" || typeof encoder !== "function") {
    io.stderr("defaults exporter: provider and encoder are required\n");
    return 2;
  }

  let defaults: T;
  const profile = flags.profile as P;
  try {
    defaults = await provider(profile);
    if (defaults === null || defaults === undefined) {
      throw new TypeError("provider returned no defaults");
    }
  } catch {
    io.stderr("defaults exporter: loading profile defaults failed\n");
    return 1;
  }

  let artifact: string;
  try {
    artifact = await encoder(profile, defaults);
    if (typeof artifact !== "string") throw new TypeError("encoder returned a non-string");
    const parsed = parseDefaultsArtifact(artifact);
    if (parsed.profile !== profile) throw new TypeError("encoded profile does not match");
  } catch {
    io.stderr("defaults exporter: encoding defaults artifact failed\n");
    return 1;
  }

  try {
    if (flags.output === "-") io.stdout(artifact);
    else await writeArtifact(flags.output, artifact);
    return 0;
  } catch {
    io.stderr("defaults exporter: writing defaults artifact failed\n");
    return 1;
  }
}

interface ExporterFlags {
  readonly profile: string;
  readonly output: string;
  readonly help: boolean;
}

function parseExporterFlags(args: readonly string[]): ExporterFlags {
  const values = new Map<string, string>();
  let help = false;
  for (let index = 0; index < args.length; index += 1) {
    const argument = args[index];
    if (!argument) continue;
    if (argument === "--help" || argument === "-h") {
      help = true;
      continue;
    }
    if (!argument.startsWith("--")) {
      throw new TypeError("defaults exporter: positional arguments are not supported");
    }
    const separator = argument.indexOf("=");
    const name = separator === -1 ? argument : argument.slice(0, separator);
    const inline = separator === -1 ? undefined : argument.slice(separator + 1);
    if (name !== "--profile" && name !== "--output") {
      throw new TypeError(`defaults exporter: unknown option ${boundedOption(name)}`);
    }
    if (values.has(name)) throw new TypeError(`defaults exporter: duplicate option ${name}`);
    const value = inline ?? args[index + 1];
    if (inline === undefined) index += 1;
    if (!value || (inline === undefined && value.startsWith("--"))) {
      throw new TypeError(`defaults exporter: ${name} requires a value`);
    }
    values.set(name, value);
  }
  if (help) return { profile: "", output: "", help: true };
  const profile = values.get("--profile");
  if (!profile) throw new TypeError("defaults exporter: --profile is required");
  validateDefaultsProfile(profile);
  const output = values.get("--output") ?? "-";
  if (output.trim() !== output) {
    throw new TypeError("defaults exporter: --output must be canonical");
  }
  return { profile, output, help: false };
}

async function writeArtifact(path: string, artifact: string): Promise<void> {
  const destination = resolve(path);
  const directory = dirname(destination);
  await mkdir(directory, { recursive: true, mode: 0o755 });
  const temporary = `${directory}/.kms-defaults-${process.pid}-${randomUUID()}`;
  let mode = 0o600;
  try {
    mode = (await stat(destination)).mode & 0o777;
  } catch (error) {
    if (!isNotFound(error)) throw error;
  }
  let committed = false;
  try {
    const handle = await open(temporary, "wx", mode);
    try {
      await handle.writeFile(artifact, "utf8");
      await handle.sync();
    } finally {
      await handle.close();
    }
    await rename(temporary, destination);
    committed = true;
  } finally {
    if (!committed) {
      try {
        await unlink(temporary);
      } catch {
        // Preserve the primary write error.
      }
    }
  }
}

function boundedMessage(error: unknown, fallback: string): string {
  const message = error instanceof Error ? error.message : fallback;
  return Buffer.from(message.replace(/[\r\n\t]/gu, " "), "utf8")
    .subarray(0, MAX_ERROR_BYTES)
    .toString("utf8");
}

function boundedOption(option: string): string {
  return JSON.stringify(option.slice(0, 80));
}

function isNotFound(error: unknown): boolean {
  return (
    typeof error === "object" &&
    error !== null &&
    "code" in error &&
    (error as { code?: unknown }).code === "ENOENT"
  );
}

function exporterUsage(): string {
  return `Usage: defaults-exporter --profile <profile> [--output <file|->]\n\nOptions:\n  --profile <profile>  Application defaults profile\n  --output <file|->    Artifact destination; defaults to stdout\n  --help               Show this help\n`;
}
